src/plugins/cookies/index.js

```diff
@@ -0,0 +1,150 @@
+        assertGet.apply(null, arguments);
+        const params = {
+            name: details.name,
+            domain: new URI(details.url).hostname()
+        };
+        return this._invokeResult(_.findWhere(this._state, params) || null, callback);
+    }
+
+    /**
+     * get all cookie list by criteria
+     * @param {AllCookieCriteria} details
+     * @param {Function} callback
+     */
+    getAll (details, callback) {
+        assertGetAll.apply(this, arguments);
+        const params = details;
+        if (params.url) {
+            params.domain = new URI(details.url).hostname();
+            delete params.url;
+        }
+        return this._invokeResult(_.where(this._state, params), callback);
+    }
+
+    /**
+     * set cookie value
+     * @param {ChromeCookie} details
+     * @param {Function} callback
+     */
+    set (details, callback) {
+        assertSet.apply(null, arguments);
+        const cookie = new ChromeCookie(details);
+        const cookieInfo = cookie.toString();
+        this._appendCookie(cookieInfo);
+        this._invokeResult(cookieInfo, callback);
+    }
+
+    /**
+     * remove cookie
+     * @param {Object} details
+     * @param {String} details.url
+     * @param {String} details.name
+     * @param {Function} [callback]
+     */
+    remove (details, callback) {
+        assertRemove.apply(null, arguments);
+        const params = {
+            name: details.name,
+            domain: (new URI(details.url)).hostname()
+        };
+        const cookieInfo = _.findWhere(this._state, params);
+        if (cookieInfo) {
+            const index = _.findIndex(this._state, cookieInfo);
+            this._state.splice(index, 1);
+            this._triggerChange({cause: 'explicit', removed: true, cookie: cookieInfo});
+        }
+        this._invokeResult(details, callback);
+    }
+
+    /**
+     * Append new cookie
+     * @param {Object} cookieInfo
+     * @private
+     */
+    _appendCookie (cookieInfo) {
+        const index = _.findIndex(this._state, {
+            name: cookieInfo.name,
+            domain: cookieInfo.domain
+        });
+        if (index >= 0) {
+            this._state.splice(index, 1, cookieInfo);
+            this._triggerChange({cause: 'overwrite', removed: true, cookie: cookieInfo});
+            this._triggerChange({cause: 'explicit', removed: false, cookie: cookieInfo});
+        } else {
+            this._state.push(cookieInfo);
+            this._triggerChange({cause: 'explicit', removed: false, cookie: cookieInfo});
+        }
+    }
+
+    /**
+     * Trigger change event
+     * @param {Object} changeInfo
+     * @private
+     */
+    _triggerChange (changeInfo) {
+        this.onChanged.triggerAsync(changeInfo);
+    }
+
+    /**
+     * Async invoke result
+     * @param {*} result
+     * @param {Function} callback
+     * @private
+     */
+    _invokeResult (result, callback) {
+        setTimeout(() => {
+            if (_.isFunction(callback)) {
```

Comment:
You can check this condition before the timeout setting to avoid a redundant timeout trigger.